### DISCUSSION TITLE
✔ Where is the structured header reference hosted?

### DISCUSSION THREAD

**Discussion by ecki** [oshp/oshp-tracking#9] [Q&A] [answered]
While watching some presentations of the project I got the impression there is somewhere a canonical place for a structured list of headers. But it’s not really mentioned anywhere. The REST API design document also mentions a canonical source for all headers. So where is it? :)

i especially would like to have the list of headers to avoid in a self refreshing url (and not sure I need a rest service for it, but if you want to provide it it’s also fine, where will it be available?)

**Comment by righettod** [comment]
Hi,

You remarks are totally accurate and I share them totally 😃 

During the work on the API, I was figuring that we need to have a way to store the reference headers configuration in a machine processable way.

💬 I can propose this idea

Replace the following tables:

![image](https://user-images.githubusercontent.com/1573775/186580320-7005ab7b-e87f-429d-b281-378d97d8f2b8.png)

![image](https://user-images.githubusercontent.com/1573775/186580352-c7d3da73-d2c0-454d-a9da-eed2923dd751.png)

By a JSON file stored in the [website GH repo](https://github.com/OWASP/www-project-secure-headers/) using such content:

📑 [Link to JSON Viewer](https://jsonvisio.com/editor?json=%5B%5B%22add%22%2C%22remove%22%2C%22a%7C0%7C1%22%2C%22name%22%2C%22value%22%2C%22a%7C3%7C4%22%2C%22Strict%20Transport%20Security%22%2C%22max-age%3D31536000%20%3B%20includeSubDomains%22%2C%22o%7C5%7C6%7C7%22%2C%22X-Frame-Options%22%2C%22deny%22%2C%22o%7C5%7C9%7CA%22%2C%22a%7C8%7CB%22%2C%22Server%22%2C%22Liferay-Portal%22%2C%22a%7CD%7CE%22%2C%22o%7C2%7CC%7CF%22%5D%2C%22G%22%5D)

![image](https://user-images.githubusercontent.com/1573775/186581257-2836ec24-f2b5-4cd4-88d8-5f44234a75c4.png)


```json
{
  "add": [
    {
      "name": "Strict Transport Security",
      "value": "max-age=31536000 ; includeSubDomains"
    },
    {
      "name": "X-Frame-Options",
      "value": "deny"
    }
  ],
  "remove": [
    "Server",
    "Liferay-Portal"
  ]
}
```

@riramar @ecki  What do you think?

**Comment by riramar** [comment]
@righettod I agree with your aproach.

**Comment by ecki** [comment]
Yes, one or possibly two JSON files for those usecases. I would go with two as the Schema are quite different. But the question is, how to generate the readable tables on the web site out of it? (For the remove well have a „description“ and „example“ column?)

**Comment by righettod** [comment]
> Yes, one or possibly two JSON files for those usecases. I would go with two as the Schema are quite different. 

Make sense, let's go for 2 JSON files.

> But the question is, how to generate the readable tables on the web site out of it? (For the remove well have a „description“ and „example“ column?)

Perhaps I'm wrong, but, the way in which the HTML site is generated will cause some issue if we generate the markdown content dynamically. 

My proposal to follow your proposal is to add HTML comment as marker for both tables on the markdown file and then have a GitHub action that generate the both JSON file based on markdown content automatically when the markdown is updated. With python it is possible in a stable way.

**Comment by righettod** [comment]
Implemented with https://github.com/OWASP/www-project-secure-headers/pull/88

**Reply by righettod** [reply]
I submitted the https://github.com/OWASP/www-project-secure-headers/pull/89 to try to fix the broken tables 😥

**Comment by righettod** [comment, marked as answer]
✅ References files now provided and automatically updated:
* [Headers to add](https://github.com/OWASP/www-project-secure-headers/blob/master/ci/headers_add.json)
* [Headers to remove](https://github.com/OWASP/www-project-secure-headers/blob/master/ci/headers_remove.json)

⚒ They will be used, as foundation, for the work on #6
